### PR TITLE
Use `MockedClientHandle` in other tests

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

While adding a field to the `Client` type, I realized I would have to add a dummy instance of the field to many mock client instances in tests. Since I had written a `MockedClientHandle` for some other tests, I wondered if I could use that in the other tests so that I would only need to update one place later.

## Solution

I moved the `MockedClientHandle` to another module to make it easier to find, and also extended it to perform the operations the other tests required. I ended up trying to refactor those tests a bit to try to improve readability.

## Review

@teor2345 recently wrote some of the changed tests.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-network/src/peer/client/tests/vectors.rs [hunk 1]
```diff
@@ -1,233 +1,102 @@
 //! Fixed peer [`Client`] test vectors.
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests/vectors.rs:1]: Coverage Reduction: The changes in this file produce good test coverage for `MockedClientHandle`, but reduce coverage for `peer::Client`.

For example, after these changes, we don't have any tests that the real `peer::Client` is cancellation-safe.

If we want to make sure that `MockedClientHandle` behaves like `peer::Client`, let's copy the tests instead. Otherwise, let's revert these changes.

File: zebra-network/src/peer_set/set/tests.rs [hunk 2]
```diff
@@ -21,10 +18,7 @@ use zebra_chain::{
 
 use super::MorePeers;
 use crate::{
-    peer::{
-        CancelHeartbeatTask, Client, ClientRequest, ErrorSlot, LoadTrackedClient,
-        MinimumPeerVersion,
-    },
+    peer::{LoadTrackedClient, MinimumPeerVersion, MockedClientHandle},
```
Comment (teor2345) [inline zebra-network/src/peer_set/set/tests.rs]: Question: How do we know when we can get accurate results from testing `PeerSet<MockedClientHandle>`? How do we know when we need to test `PeerSet<peer::Client>` for good test results?

Where are the tests that make sure that `MockedClientHandle` behaves like `peer::Client`? How much is it ok for the implementations to diverge?

Can you add some documentation explaining when to use each of them?

File: zebra-network/src/peer_set/initialize/tests/vectors.rs [hunk 3]
```diff
@@ -351,16 +348,7 @@ async fn crawler_peer_limit_one_connect_ok_then_drop() {
...
-            let (shutdown_tx, _shutdown_rx) = oneshot::channel();
-            let error_slot = ErrorSlot::default();
-
-            let fake_client = peer::Client {
-                shutdown_tx: Some(shutdown_tx),
-                server_tx,
-                error_slot,
-                version: Version(1),
-            };
+            let (_handle, fake_client) = MockedClientHandle::new(Version(1));
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize/tests/vectors.rs]: Possible Coverage Reduction: How can we be sure that these changes still produce good coverage for `peer::Client` and the zebra-network async tasks?

It's probably ok, because the `PeerSet` is fake anyway, and we're just counting peers. But let's document that.

File: zebra-test/src/service_extensions.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+//! Extension traits for [`Service`] types to help with testing.
+
+use futures::FutureExt;
+use tower::{Service, ServiceExt};
+
+/// An extension trait to check if a [`Service`] is immediately ready to be called.
+pub trait IsReady<Request>: Service<Request> {
```
Comment (teor2345) [inline zebra-test/src/service_extensions.rs:9]: Undocumented Behaviour: this convenience wrapper hides a call to `now_or_never`, which never registers the task for waking. So after these methods are called once on a service, any further calls to the service may produce incorrect test results.

So we can either:
1. Revert this change, or
2. Make the new wrapper harder to misuse:
    * add `OrNever` to the wrapper name
    * modify its methods to take ownership of the service
    * summarise the `now_or_never` behaviour in the docs

It would also be nice to prevent the wrapper methods being called on a cloned, buffered service. But our only way of doing that is educating developers. We can teach everyone about the wrapper, or revert the wrapper, and just teach about `now_or_never`.

We can't switch to a `now_or_never` alternative that can be called multiple times. Because we can't obtain service readiness multiple times anyway. (Readiness must be used after it is obtained, particularly for buffered services.)

**Review by teor2345 — DISMISSED** [review]
This is a good idea, ~but we have to be careful that we don't reduce actual `peer::Client` test coverage.~

File: zebra-network/src/peer/client/tests/vectors.rs (hunk 1 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests/vectors.rs:1]: I don't understand. The coverage should be exactly the same. The `MockedClientHandle` is just a holder of all the endpoints to channels sent into the `Client` instance.

I only call it a mocked client because there is no actual outside connection.

File: zebra-network/src/peer_set/set/tests.rs (hunk 2 above)
Comment (jvff) [reply zebra-network/src/peer_set/set/tests.rs]: It should be exactly the same. `MockedClientHandle::new()` returns `(Self, Client)`, and the `Client` is what is passed to the `PeerSet`. The handle is used just to access things inside the `Client`.

File: zebra-test/src/service_extensions.rs (hunk 4 above)
Comment (jvff) [reply zebra-test/src/service_extensions.rs:9]: This is a valid concern, and I thought it might be a good idea to use `NowOrLater` instead. To do so, I had to move the `NowOrLater` type from `zebrad` into a new `now-or-later` crate in order to avoid cyclic dependencies.

File: now-or-later/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "now-or-later"
+version = "1.0.0-beta.2"
```
Comment (teor2345) [inline now-or-later/Cargo.toml]: Since this is a non-zebra crate, it should have its own version numbers:
```suggestion
version = "0.0.1"
```

File: Cargo.toml [hunk 6]
```diff
@@ -12,6 +12,7 @@ members = [
         "zebra-utils",
         "tower-batch",
         "tower-fallback",
+        "now-or-later",
```
Comment (teor2345) [inline Cargo.toml]: Can you update the release PR template to list this new crate?

File: zebra-network/src/peer/client/tests/vectors.rs (hunk 1 above)
Comment (teor2345) [reply zebra-network/src/peer/client/tests/vectors.rs:1]: Oops, this was my mistake.

I opened PR #3246 so we get coverage comments on PRs.

Can you please improve the documentation for `MockedClientHandle`? And maybe change the name?

Specifically, this is unclear:
> A handle to a mocked [`Client`] instance.

I thought it meant that we'd replaced the functionality of the `Client` struct with a test-only implementation. But we're actually only mocking the channels connected to it. So maybe it's a `ClientWithMockedChannels` or `ClientTestHarness`?

File: zebra-test/src/service_extensions.rs (hunk 4 above)
Comment (teor2345) [reply zebra-test/src/service_extensions.rs:9]: Thanks for this fix!

How can we prevent multiple service readiness calls in the tests?

> We can't switch to a now_or_never alternative that can be called multiple times. Because we can't obtain service readiness multiple times anyway. (Readiness must be used after it is obtained, particularly for buffered services.)

If the service isn't buffered, and it has no internal reservations, it is ok to check readiness multiple times:
https://docs.rs/tower/0.4.11/tower/trait.Service.html#tymethod.poll_ready

But if the service makes internal reservations, or it is buffered, or it checks readiness for buffered services, we can't call it multiple times:
https://docs.rs/tower/0.4.11/tower/buffer/struct.Buffer.html#a-note-on-choosing-a-bound

I'd like to take ownership of the service, if we can. And I'd like to document this issue in `IsReady`.

File: zebra-network/src/peer/client/tests.rs [hunk 7]
```diff
@@ -1,3 +1,153 @@
...
+    }
+
+    /// Gets the peer protocol version associated to the [`Client`].
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    /// Checks if the [`Client`] instance has not been dropped, which would have disconnected from
+    /// the peer.
+    pub fn is_connected(&mut self) -> bool {
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]: The `peer::Client` is not directly connected to the peer.

Instead, the peer TCP connection is split into `Connection.peer_tx` and `Connection::run(peer_rx)`. It is disconnected when they are both dropped.

```suggestion
    /// Returns true if the [`Client`] instance still wants connection heartbeats to be sent.
    ///
    /// Checks that the client:
    /// - has not been dropped,
    /// - has not closed or dropped the mocked heartbeat task channel, and
    /// - has not asked the mocked heartbeat task to shut down.
    pub fn wants_connection_heartbeats(&mut self) -> bool {
```

File: zebra-network/src/peer/client/tests.rs [hunk 8]
```diff
@@ -1,3 +1,153 @@
...
+            .try_recv();
+
+        match receive_result {
+            Ok(None) => true,
+            Ok(Some(CancelHeartbeatTask)) | Err(oneshot::Canceled) => false,
+        }
+    }
+
+    /// Drops the shutdown receiver endpoint.
+    pub fn drop_shutdown_receiver(&mut self) {
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// Drops the mocked heartbeat shutdown receiver endpoint.
    pub fn drop_heartbeat_shutdown_receiver(&mut self) {
```

File: zebra-network/src/peer/client/tests.rs [hunk 9]
```diff
@@ -1,3 +1,153 @@
...
+        let _ = self
+            .shutdown_receiver
+            .take()
+            .expect("request receiver endpoint has already been dropped");
+    }
+
+    /// Closes the request receiver endpoint.
+    pub fn close_request_receiver(&mut self) {
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// Closes the outbound peer request receiver endpoint.
    pub fn close_outbound_peer_request_receiver(&mut self) {
```

File: zebra-network/src/peer/client/tests.rs [hunk 10]
```diff
@@ -1,3 +1,153 @@
...
+        self.request_receiver
+            .as_mut()
+            .expect("request receiver endpoint has been dropped")
+            .close();
+    }
+
+    /// Drops the request receiver endpoint, forcefully closing the channel.
+    pub fn drop_request_receiver(&mut self) {
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// Drops the outbound peer request receiver endpoint, forcefully closing the channel.
    pub fn drop_outbound_peer_request_receiver(&mut self) {
```

File: zebra-network/src/peer/client/tests.rs [hunk 11]
```diff
@@ -1,3 +1,153 @@
...
+/// A handle to a mocked [`Client`] instance.
+pub struct MockedClientHandle {
+    request_receiver: Option<mpsc::Receiver<ClientRequest>>,
+    shutdown_receiver: Option<oneshot::Receiver<CancelHeartbeatTask>>,
+    error_slot: ErrorSlot,
+    version: Version,
+}
+
+impl MockedClientHandle {
+    /// Create a new mocked [`Client`] instance, returning it together with a handle to track it.
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// Create a new [`Client`] instance with mocked channels, returning it together with a handle to track it.
```

File: zebra-network/src/peer/client/tests.rs [hunk 12]
```diff
@@ -1,3 +1,153 @@
...
 mod vectors;
+
+use futures::channel::{mpsc, oneshot};
+
+use crate::{
+    peer::{error::SharedPeerError, CancelHeartbeatTask, Client, ClientRequest, ErrorSlot},
+    protocol::external::types::Version,
+};
+
+/// A handle to a mocked [`Client`] instance.
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
/// A handle to the mocked channels for testing a [`Client`] instance.
```

File: zebra-network/src/peer/client/tests.rs [hunk 13]
```diff
@@ -1,3 +1,153 @@
...
+
+    /// Drops the request receiver endpoint, forcefully closing the channel.
+    pub fn drop_request_receiver(&mut self) {
+        self.request_receiver
+            .take()
+            .expect("request receiver endpoint has already been dropped");
+    }
+
+    /// Tries to receive a [`ClientRequest`] sent by the mocked [`Client`] instance.
+    pub(crate) fn try_to_receive_request(&mut self) -> ReceiveRequestAttempt {
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// Tries to receive a [`ClientRequest`] sent by the [`Client`] instance.
    /// The outbound peer that would receive this request is mocked for testing.
    pub(crate) fn try_to_receive_outbound_peer_request(&mut self) -> ReceiveRequestAttempt {
```

File: zebra-network/src/peer/client/tests.rs [hunk 14]
```diff
@@ -1,3 +1,153 @@
...
+    /// If there's already an error in the [`ErrorSlot`].
+    pub fn set_error(&self, error: impl Into<SharedPeerError>) {
+        self.error_slot
+            .try_update_error(error.into())
+            .expect("unexpected earlier error in error slot")
+    }
+}
+
+/// A representation of the result of an attempt to receive a [`ClientRequest`] sent by the mocked
+/// [`Client`] instance.
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
/// The result of an attempt to receive a [`ClientRequest`] sent by a [`Client`] instance.
/// The outbound peer that would receive this request is mocked for testing.
```

File: zebra-network/src/peer/client/tests.rs [hunk 15]
```diff
@@ -1,3 +1,153 @@
...
+pub(crate) enum ReceiveRequestAttempt {
+    /// The mocked [`Client`] instance has closed the sender endpoint of the channel.
```
Comment (teor2345) [inline zebra-network/src/peer/client/tests.rs]:
```suggestion
    /// The [`Client`] instance has closed the sender endpoint of the mocked channel.
```

File: zebra-test/src/service_extensions.rs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    /// Check if the [`Service`] is immediately ready to be called.
+    fn is_ready(&mut self) -> BoxFuture<bool>;
+
+    /// Check if the [`Service`] is not immediately ready because it returns an error.
+    fn not_ready_due_to_error(&mut self) -> BoxFuture<bool>;
+}
+
+impl<S, Request> IsReady<Request> for S
```
Comment (teor2345) [inline zebra-test/src/service_extensions.rs:20]: Can you also add a method that explicitly checks `is_pending`?

File: zebra-test/src/service_extensions.rs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    S: Service<Request> + Send,
+    Request: 'static,
+{
+    fn is_ready(&mut self) -> BoxFuture<bool> {
+        NowOrLater(self.ready())
+            .map(|ready_result| matches!(ready_result, Some(Ok(_))))
+            .boxed()
+    }
+
+    fn not_ready_due_to_error(&mut self) -> BoxFuture<bool> {
```
Comment (teor2345) [inline zebra-test/src/service_extensions.rs]: Naming: Can we try to avoid negated names, particularly for boolean functions?

This is consistent with the `Service::poll_ready` contract:
> If Poll::Ready(Err(_)) is returned, the service is no longer able to service requests and the caller should discard the service instance.

https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html

```suggestion
    fn is_failed(&mut self) -> BoxFuture<bool> {
```

**Comment by jvff** [comment]
I changed how the initialization is done to use a `MockClientBuilder`, in order to:

1. Try to make it clearer that a real `Client` instance is created
2. `MockedClientHandle` only contains the mocked data fed into the `Client`
3. Prepare for writing tests for tracking the background tasks. I'll need the builder pattern to be able to configure custom mock background tasks.

**Review by teor2345 — COMMENTED** [review]
I'm sorry for the misunderstandings in my last review, I found these changes really confusing. 

I have a bunch of suggestions for improving the documentation and names in this PR.

We might also want to improve the corresponding `peer::Client` and `peer::Connection` documentation, because it is outdated.

File: zebra-network/src/peer/client/tests.rs (hunk 9 above)
Comment (teor2345) [reply zebra-network/src/peer/client/tests.rs]: Actually, maybe this is unclear - what I mean is "the receiver endpoint for outbound requests to the mocked peer".

File: now-or-later/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+authors = ["Zcash Foundation <zebra@zfnd.org>"]
+license = "MIT OR Apache-2.0"
+edition = "2018"
+
+[dependencies]
+pin-project = "1.0.7"
+
+[dev-dependencies]
+futures = "0.3.18"
+tokio = { version = "1.14.0", features = ["full"] }
```
Comment (teor2345) [inline now-or-later/Cargo.toml]: - https://github.com/ZcashFoundation/zebra/pull/3245

```suggestion
tokio = { version = "1.15.0", features = ["full"] }
```

**Comment by teor2345** [comment]
This is a low priority by itself, but #3253 depends on it, so I've marked it as medium.

File: zebra-network/src/peer/client/tests/vectors.rs (hunk 1 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests/vectors.rs:1]: Done:

- `MockClientBuilder` -> `ClientTestHarnessBuilder`
- `MockedClientHandle` -> `ClientTestHarness`

Let me know what you think.

File: zebra-test/src/service_extensions.rs (hunk 4 above)
Comment (jvff) [reply zebra-test/src/service_extensions.rs:9]: This seems like non-trivial work. Could you open an issue for us to do it later?

File: now-or-later/Cargo.toml (hunk 5 above)
Comment (jvff) [reply now-or-later/Cargo.toml]: Done.

File: Cargo.toml (hunk 6 above)
Comment (jvff) [reply Cargo.toml]: Done.

File: zebra-network/src/peer/client/tests.rs (hunk 9 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests.rs]: Done, I tried to phrase it to be clearer. Let me know what you think.

File: zebra-network/src/peer/client/tests.rs (hunk 10 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests.rs]: Done, I tried to phrase it to be clearer. Let me know what you think.

File: zebra-network/src/peer/client/tests.rs (hunk 11 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests.rs]: This constructor has been removed and replaced with the `TestClientBuilder`.

File: zebra-network/src/peer/client/tests.rs (hunk 13 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests.rs]: Done, I tried to phrase it to be clearer. Let me know what you think.

File: zebra-network/src/peer/client/tests.rs (hunk 14 above)
Comment (jvff) [reply zebra-network/src/peer/client/tests.rs]: Done, I tried to phrase it to be clearer. Let me know what you think.

File: zebra-test/src/service_extensions.rs (hunk 16 above)
Comment (jvff) [reply zebra-test/src/service_extensions.rs:20]: Done.

File: now-or-later/Cargo.toml (hunk 18 above)
Comment (jvff) [reply now-or-later/Cargo.toml]: Done.

File: zebra-test/src/service_extensions.rs (hunk 4 above)
Comment (teor2345) [reply zebra-test/src/service_extensions.rs:9]: Since `mpsc` channels have capacity limits, this could be a real bug in this refactor.

Allowing the service to be re-used after `is_ready` could hide test regressions in this PR. So I'd really like to do it as part of this PR, to make sure the refactor works correctly.

The previous code took ownership of the `ServiceExt::ready` future, but didn't take ownership of the service. So the bug could still have happened then. (But I was very careful to avoid it.)

I would like to make it easier to maintain this code, and avoid the risk of subtle bugs in this PR, or in future refactors. I'll see if I can fix it today.

Comment (teor2345) [reply zebra-test/src/service_extensions.rs:9]: > Could you open an issue for us to do it later?

See PR #3262, which takes ownership of the service, and documents this issue in `IsReady`.

I also opened ticket #3261 for some larger API changes.

Comment (teor2345) [reply zebra-test/src/service_extensions.rs:9]: Here's an example of a potential bug discovered by this fix:

> assertion failed: harness.current_error().is_none()

in `client_service_ready_ok`:

```rust
    assert!(client.is_ready().await);
    assert!(harness.current_error().is_none());
    assert!(harness.wants_connection_heartbeats());
    assert!(harness.try_to_receive_outbound_client_request().is_empty());
```

https://github.com/ZcashFoundation/zebra/runs/4577397601?check_suite_focus=true#step:14:984

`client.is_ready()` might reserve capacity in the mpsc channel (it's undocumented), and the channel is small, so we need to document why the test is valid. Or re-arrange the test so it doesn't matter.

I covered both scenarios in PR #3262 by adding a new test (just in case the behaviour is ever different).

**Review by teor2345 — DISMISSED** [review]
I'm happy with this PR, ~~but I'd like to merge the fixes in PR #3262 soon as well.~~

**Review by teor2345 — COMMENTED** [review]
In https://github.com/ZcashFoundation/zebra/pull/3262#issuecomment-997954535 @jvff said:

> Doing the assertion last is also not ideal, since some assertions require the client instance to be polled first.

Do these requirements apply to all `Client`s, or only to `ClientTestHarness`?

This isn't a blocker, but it would be nice to document these requirements as part of this PR. (They were a bit of a surprise to me.)

**Comment by teor2345** [comment]
> > Doing the assertion last is also not ideal, since some assertions require the client instance to be polled first.
> 
> Do these requirements apply to all `Client`s, or only to `ClientTestHarness`?
> 
> This isn't a blocker, but it would be nice to document these requirements as part of this PR. (They were a bit of a surprise to me.)

PR #3273 might help clarify these requirements. If I'm using the `Client` or test harness wrong, let me know, and I'll fix it.

**Comment by jvff** [comment]
> > Doing the assertion last is also not ideal, since some assertions require the client instance to be polled first.
> 
> 
> Do these requirements apply to all `Client`s, or only to `ClientTestHarness`?

Sorry I wasn't clear. I meant that doing it last would mean that we wouldn't be asserting on the correct things. We would assert things before `client.is_ready().await`, and what we actually want is to assert that things changed after the call to `client.is_ready().await`.

> PR #3273 might help clarify these requirements. If I'm using the `Client` or test harness wrong, let me know, and I'll fix it.

Looks good to me :+1:

**Review by teor2345 — APPROVED** [review]
I think we're all done here!

**Comment by teor2345** [comment]
Note for changelog: this PR includes 3 changes:

Use `MockedClientHandle` in other tests (#3241)
Add extra client tests for zero and multiple readiness checks (#3273) 
Replace `NowOrLater` with `futures::poll!` (#3272)
